lib/space_saving.c

```diff
@@ -0,0 +1,795 @@
+			RTE_LOG(ERR, HASH, "Bucket number %d could not be "
+				"created!\n",bkt_id);
+			ret = -1;
+		}
+	}
+	return ret;
+}
+
+/* Retuns a pointer to a counter bucket */
+static int
+get_bucket(int bkt_id, int proto, struct counter_bucket *ct_bkt)
+{
+	int ret;
+	char bkt_name[64];
+
+	ret = snprintf(bkt_name, sizeof(bkt_name),
+		"Bucket_number_%d_%d", proto, bkt_id);
+	RTE_VERIFY(ret > 0 && ret < (int)sizeof(bkt_name));
+
+	ct_bkt->proto = proto;
+
+	switch(proto) {
+		case ETHER_TYPE_IPv4:
+			ct_bkt->bkt.bkt_ip4 = rte_hash_find_existing(bkt_name);
+			if(ct_bkt->bkt.bkt_ip4 == NULL) {
+				RTE_LOG(ERR, HASH, "Bucket number %d is not "
+					"present!\n", bkt_id);
+				ret = -2;
+			}
+			break;
+
+		case ETHER_TYPE_IPv6:
+			ct_bkt->bkt.bkt_ip6 = rte_hash_find_existing(bkt_name);
+			if(ct_bkt->bkt.bkt_ip6 == NULL) {
+				RTE_LOG(ERR, HASH, "Bucket number %d is not "
+					"present!\n", bkt_id);
+				ret = -2;
+			}
+			break;
+
+		case ETHER_TYPE_ARP:
+			ret = -1;
+			break;
+
+		default:
+			RTE_LOG(ERR, HASH, "Unknown network layer protocol\n");
+			ret = -1;
+			break;
+	}
+	return ret;
+}
+
+/* Returns data from a counter bucket */
+static int
+get_bucket_entry(struct counter_bucket *ct_bkt, struct ip_flow *flow,
+			struct bucket_data *bkt_data)
+{
+	int ret;
+
+	switch(flow->proto) {
+		case ETHER_TYPE_IPv4:
+			ret = rte_hash_lookup_data(ct_bkt->bkt.bkt_ip4, flow,
+							(void **)bkt_data);
+			if(ret < 0) {
+				RTE_LOG(ERR, HASH, "Bucket entry is not "
+					"present!\n");
+				ret = -1;
+			}
+			break;
+
+		case ETHER_TYPE_IPv6:
+			ret = rte_hash_lookup_data(ct_bkt->bkt.bkt_ip6, flow,
+							(void **)bkt_data);
+			if(ret < 0) {
+				RTE_LOG(ERR, HASH, "Bucket entry is not "
+					"present!\n");
+				ret = -1;
+			}
+			break;
+
+		case ETHER_TYPE_ARP:
+			ret = -1;
+			break;
+
+		default:
+			RTE_LOG(ERR, HASH, "Unknown network layer protocol\n");
+			ret = -1;
+		break;
+	}
+	return ret;
+}
+
+/* Add data to a counter bucket */
+static int
+add_bucket_entry(struct counter_bucket *ct_bkt, struct ip_flow *flow,
+			struct bucket_data *bkt_data)
+{
+	int ret;
+
+	switch(flow->proto) {
+		case ETHER_TYPE_IPv4:
+			ret = rte_hash_add_key_data(ct_bkt->bkt.bkt_ip4,
+							flow, bkt_data);
+			if(ret < 0) {
+				RTE_LOG(ERR, HASH, "Data cannot be added into "
+					"the bucket\n");
+				ret = -1;
+			}
+			break;
+
+		case ETHER_TYPE_IPv6:
+			ret = rte_hash_add_key_data(ct_bkt->bkt.bkt_ip6,
+							flow, bkt_data);
+			if(ret < 0) {
+				RTE_LOG(ERR, HASH, "Data cannot be added into " 
+					"the bucket\n");
+				ret = -1;
+			}
+			break;
+
+		case ETHER_TYPE_ARP:
+			ret = -1;
+			break;
+
+		default:
+			RTE_LOG(ERR, HASH,"Unknown network layer protocol!\n");
+			ret = -1;
+			break;
+	}
+	return ret;
+}
+
+/* Deletes data from a counter bucket */
+static int
+delete_bucket_entry(struct counter_bucket *ct_bkt, struct ip_flow *flow)
+{
+	int ret;
+
+	switch(flow->proto) {
+		case ETHER_TYPE_IPv4:
```

Comment:
Same thing here: add a space and bring everything inside of this `switch` back one level of indentation.